### PR TITLE
Support different Database types for Query optimization and fix some bugs

### PR DESCRIPTION
**Pull request by JiekerTime**
about issue https://github.com/apache/shardingsphere/issues/8880

### CODE REVIEW

File: shardingsphere-infra/shardingsphere-infra-optimize/src/test/java/org/apache/shardingsphere/infra/optimize/context/CalciteContextFactoryTest.java [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+        assertNotNull(logicDb);
+        Properties properties = logicDb.getConnectionProperties();
+        assertNotNull(properties);
+        assertThat(properties.getProperty("lex"), is("MYSQL"));
+        assertThat(properties.getProperty("conformance"), is("DEFAULT"));
+        CalciteLogicSchema calciteLogicSchema = logicDb.getCalciteLogicSchema();
+        assertNotNull(calciteLogicSchema);
+        assertThat(calciteLogicSchema.getName(), is("logic_db"));
+    }
+
```
Comment (tristaZero) [inline shardingsphere-infra/shardingsphere-infra-optimize/src/test/java/org/apache/shardingsphere/infra/optimize/context/CalciteContextFactoryTest.java:64]: Please remove this redundant blank line.

File: shardingsphere-infra/shardingsphere-infra-optimize/src/test/java/org/apache/shardingsphere/infra/optimize/context/CalciteContextFactoryTest.java [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Properties;
+
+import static org.hamcrest.CoreMatchers.is;
+import static org.junit.Assert.assertNotNull;
+import static org.junit.Assert.assertThat;
+
+public class CalciteContextFactoryTest {
```
Comment (tristaZero) [inline shardingsphere-infra/shardingsphere-infra-optimize/src/test/java/org/apache/shardingsphere/infra/optimize/context/CalciteContextFactoryTest.java:40]: Define it as final class plz

File: shardingsphere-infra/shardingsphere-infra-optimize/src/main/java/org/apache/shardingsphere/infra/optimize/schema/CalciteLogicSchemaFactory.java [hunk 3]
```diff
@@ -44,7 +44,7 @@ public CalciteLogicSchemaFactory(final Map<String, ShardingSphereMetaData> metaD
             }
         }
     }
-    
+
```
Comment (tristaZero) [inline shardingsphere-infra/shardingsphere-infra-optimize/src/main/java/org/apache/shardingsphere/infra/optimize/schema/CalciteLogicSchemaFactory.java:47]: Please check style for these new blank lines.

File: shardingsphere-infra/shardingsphere-infra-optimize/src/main/java/org/apache/shardingsphere/infra/optimize/context/CalciteContextFactory.java [hunk 4]
```diff
@@ -54,68 +63,106 @@
...
         PlannerInitializer.init(planner);
         return RelOptCluster.create(planner, new RexBuilder(typeFactory));
     }
-    
+
     /**
      * Create.
      *
-     * @param schema schema
+     * @param schema  schema
```
Comment (tristaZero) [inline shardingsphere-infra/shardingsphere-infra-optimize/src/main/java/org/apache/shardingsphere/infra/optimize/context/CalciteContextFactory.java:149]: Why do you add a space?

File: shardingsphere-infra/shardingsphere-infra-optimize/src/main/java/org/apache/shardingsphere/infra/optimize/context/CalciteContextFactory.java [hunk 5]
```diff
@@ -54,68 +63,106 @@
...
+            this.properties.setProperty(LEX_CAMEL_NAME, Lex.MYSQL.name());
+            this.properties.setProperty(CONFORMANCE_CAMEL_NAME, SqlConformanceEnum.STRICT_92.name());
+            return;
+        }
+        if (databaseType instanceof SQLServerDatabaseType) {
+            this.properties.setProperty(LEX_CAMEL_NAME, Lex.SQL_SERVER.name());
+            this.properties.setProperty(CONFORMANCE_CAMEL_NAME, SqlConformanceEnum.SQL_SERVER_2008.name());
+            return;
+        }
+        throw new ShardingSphereException("No matching DatabaseType found");
```
Comment (tristaZero) [inline shardingsphere-infra/shardingsphere-infra-optimize/src/main/java/org/apache/shardingsphere/infra/optimize/context/CalciteContextFactory.java:137]: I guess there is no any chance running here, right? Please consider removing it.

File: shardingsphere-infra/shardingsphere-infra-optimize/src/main/java/org/apache/shardingsphere/infra/optimize/context/CalciteContextFactory.java [hunk 6]
```diff
@@ -54,68 +63,106 @@
...
-    
+
     private final RelOptCluster cluster;
-    
+
     public CalciteContextFactory(final Map<String, ShardingSphereMetaData> metaDataMap) {
-        properties = createProperties();
+        factory = new CalciteLogicSchemaFactory(metaDataMap);
+        DatabaseType databaseType = metaDataMap.values().iterator().next().getResource().getDatabaseType();
+        assert databaseType != null;
```
Comment (tristaZero) [inline shardingsphere-infra/shardingsphere-infra-optimize/src/main/java/org/apache/shardingsphere/infra/optimize/context/CalciteContextFactory.java:85]: `Preconditions.checkArgument` is recommended.

**Review by tristaZero — CHANGES_REQUESTED** [review]

File: shardingsphere-infra/shardingsphere-infra-optimize/src/main/java/org/apache/shardingsphere/infra/optimize/context/CalciteContextFactory.java (hunk 5 above)
Comment (JiekerTime) [reply shardingsphere-infra/shardingsphere-infra-optimize/src/main/java/org/apache/shardingsphere/infra/optimize/context/CalciteContextFactory.java:137]: In order to maintain the robustness of the code and avoid some unknown errors, please allow me to keep this part....

**Pull request closed without merging**